### PR TITLE
Packaging pipeline now builds with PythonOp (aka running autograd.Function)

### PR DESCRIPTION
**Pull request by wschin**
This link disables UTs under training's packaging pipelines to avoid unresolved symbols defined in `libpython.so`. Those symbols are used in PythonOp and appear in `onnxruntime_providers` and `onnxruntime_framework`. It's not easy to extract them, so we made this PR for a temporal solution. Ultimately, we implement one of ideas below.
- delayed loading symbols with `dlopen` and `dlsym` in Linux
- JIT PythonOp as a custom op (like AtenOp).

Then re-enable UTs in package pipelines.

### CODE REVIEW

File: cmake/onnxruntime_session.cmake [hunk 1]
```diff
@@ -39,3 +39,6 @@ endif()
 if (onnxruntime_ENABLE_TRAINING OR onnxruntime_ENABLE_TRAINING_OPS)
   target_include_directories(onnxruntime_session PRIVATE ${ORTTRAINING_ROOT})
 endif()
+if (onnxruntime_ENABLE_TRAINING_TORCH_INTEROP)
+  onnxruntime_add_include_to_target(onnxruntime_session Python::Module) 
+endif()
```
Comment (wschin) [inline cmake/onnxruntime_session.cmake:44]:
```suggestion
endif()

```

File: cmake/CMakeLists.txt [hunk 2]
```diff
@@ -54,6 +54,7 @@ option(onnxruntime_USE_DNNL "Build with DNNL support" OFF)
 option(onnxruntime_USE_FEATURIZERS "Build ML Featurizers support" OFF)
 option(onnxruntime_DEV_MODE "Enable developer warnings and treat most of them as error." OFF)
 option(onnxruntime_BUILD_UNIT_TESTS "Build ONNXRuntime unit tests" ON)
+option(onnxruntime_DISABLE_UNIT_TESTS "Disable ONNXRuntime unit tests" OFF)
```
Comment (snnn) [inline cmake/CMakeLists.txt:57]: Would it be too confusing if we have both onnxruntime_BUILD_UNIT_TESTS and onnxruntime_DISABLE_UNIT_TESTS ?

File: tools/ci_build/build.py [hunk 3]
```diff
@@ -159,6 +159,10 @@ def parse_arguments():
         "--enable_training", action='store_true', help="Enable training in ORT.")
     parser.add_argument(
         "--enable_training_ops", action='store_true', help="Enable training ops in inference graph.")
+    parser.add_argument(
+        "--enable_training_torch_interop", action='store_true', help="Enable training kernels interop with torch.")
+    parser.add_argument(
+        "--disable_unit_tests", action='store_true', help="Disable unit tests.")
```
Comment (snnn) [inline tools/ci_build/build.py:165]: Don't we have too many such options?
```
--enable_onnx_tests=true
--enable_onnx_tests=false
--skip_onnx_tests=true
--skip_onnx_tests=false
--disable_unit_tests=true
--disable_unit_tests=false
...
```

File: tools/ci_build/build.py [hunk 4]
```diff
@@ -754,6 +758,8 @@ def generate_build_tree(cmake_path, source_dir, build_dir, cuda_home, cudnn_home
         "-Donnxruntime_ENABLE_NVTX_PROFILE=" + ("ON" if args.enable_nvtx_profile else "OFF"),
         "-Donnxruntime_ENABLE_TRAINING=" + ("ON" if args.enable_training else "OFF"),
         "-Donnxruntime_ENABLE_TRAINING_OPS=" + ("ON" if args.enable_training_ops else "OFF"),
+        "-Donnxruntime_ENABLE_TRAINING_TORCH_INTEROP=" + ("ON" if args.enable_training_torch_interop else "OFF"),
+        "-Donnxruntime_DISABLE_UNIT_TESTS=" + ("ON" if args.disable_unit_tests else "OFF"),
```
Comment (snnn) [inline tools/ci_build/build.py:762]: Some other pipelines also want to turn onnxruntime_DISABLE_UNIT_TESTS OFF. If we really need to add such an option, I suggest we make it more clear. I feel people may get confused why sometimes this option works sometimes it doesn't.

File: tools/ci_build/github/azure-pipelines/orttraining-py-packaging-pipeline-torch181-cuda102.yml [hunk 5]
```diff
@@ -3,7 +3,7 @@ trigger: none
 stages:
 - template: templates/py-packaging-training-cuda-stage.yml
   parameters:
-    build_py_parameters: --enable_training --update --build
+    build_py_parameters: --enable_training --update --build --enable_training_torch_interop --disable_unit_tests
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/orttraining-py-packaging-pipeline-torch181-cuda102.yml:6]: Please add "--cmake_extra_defines onnxruntime_BUILD_UNIT_TESTS=OFF" there. Then the changes in build.py are not needed.

Comment (wschin) [reply tools/ci_build/github/azure-pipelines/orttraining-py-packaging-pipeline-torch181-cuda102.yml:6]: Ok.

File: cmake/CMakeLists.txt (hunk 2 above)
Comment (wschin) [reply cmake/CMakeLists.txt:57]: I added `DISABLE_UNIT_TESTS` because some pipelines failed with `BUILD_UNIT_TESTS=OFF`. Let me try again. Thanks.

File: tools/ci_build/build.py (hunk 4 above)
Comment (wschin) [reply tools/ci_build/build.py]: This new flag is removed.

File: tools/ci_build/build.py (hunk 3 above)
Comment (wschin) [reply tools/ci_build/build.py]: New flag is removed.

**Comment by snnn** [comment]
If you can wait a few days, I can rework the pipeline to make it easier for you.

File: tools/ci_build/github/azure-pipelines/orttraining-py-packaging-pipeline-torch181-cuda102.yml [hunk 6]
```diff
@@ -3,7 +3,7 @@ trigger: none
 stages:
 - template: templates/py-packaging-training-cuda-stage.yml
   parameters:
-    build_py_parameters: --enable_training --update --build
+    build_py_parameters: --enable_training --update --build --enable_training_torch_interop --cmake_extra_defines onnxruntime_BUILD_UNIT_TESTS=OFF
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/orttraining-py-packaging-pipeline-torch181-cuda102.yml:6]: So this "--cmake_extra_defines onnxruntime_BUILD_UNIT_TESTS=OFF" doesn't work?

Comment (wschin) [reply tools/ci_build/github/azure-pipelines/orttraining-py-packaging-pipeline-torch181-cuda102.yml:6]: No. It doesn't. A potential fix (trying):
```
    parser.add_argument(
        "--cmake_extra_defines", action='append', nargs="+",
        help="Extra definitions to pass to CMake during build system "
        "generation. These are just CMake -D options without the leading -D.")
```
We missed `action=append`.

**Comment by snnn** [comment]
I found the problem.

From the log you can see, the actual build command was:
```
python tools/ci_build/build.py  --build_dir /build --config Release --skip_submodule_sync --parallel --build_wheel --enable_onnx_tests --enable_training --update --build --enable_training_torch_interop --cmake_extra_defines onnxruntime_BUILD_UNIT_TESTS=OFF --cmake_extra_defines CMAKE_CUDA_HOST_COMPILER=/opt/rh/devtoolset-8/root/usr/bin/cc 'CMAKE_CUDA_ARCHITECTURES=35;37;50;52;60;61;70' --use_cuda --cuda_version=10.2 --cuda_home=/usr/local/cuda-10.2 --cudnn_home=/usr/local/cuda-10.2
```

Instead of having 
```
--cmake_extra_defines A --cmake_extra_defines B
```
You should have 
```
--cmake_extra_defines A  B
```

**Comment by wschin** [comment]
> I found the problem.
> 
> From the log you can see, the actual build command was:
> 
> ```
> python tools/ci_build/build.py  --build_dir /build --config Release --skip_submodule_sync --parallel --build_wheel --enable_onnx_tests --enable_training --update --build --enable_training_torch_interop --cmake_extra_defines onnxruntime_BUILD_UNIT_TESTS=OFF --cmake_extra_defines CMAKE_CUDA_HOST_COMPILER=/opt/rh/devtoolset-8/root/usr/bin/cc 'CMAKE_CUDA_ARCHITECTURES=35;37;50;52;60;61;70' --use_cuda --cuda_version=10.2 --cuda_home=/usr/local/cuda-10.2 --cudnn_home=/usr/local/cuda-10.2
> ```
> 
> Instead of having
> 
> ```
> --cmake_extra_defines A --cmake_extra_defines B
> ```
> 
> You should have
> 
> ```
> --cmake_extra_defines A  B
> ```

We need to allow  multiple `--cmake_extra_defines` because this option are set in two places.

File: cmake/onnxruntime_training.cmake [hunk 7]
```diff
@@ -71,7 +71,7 @@ if (onnxruntime_BUILD_UNIT_TESTS)
   add_dependencies(onnxruntime_training_runner ${onnxruntime_EXTERNAL_DEPENDENCIES} onnx onnxruntime_providers)
 
   if (onnxruntime_ENABLE_TRAINING_TORCH_INTEROP)
-    target_link_libraries(onnxruntime_training_runner PRIVATE Python::Python)
+    target_link_libraries(onnxruntime_training_runner PRIVATE Python::Module)
```
Comment (pengwa) [inline cmake/onnxruntime_training.cmake:74]: changing this from ::Python to ::Module, for those training pipelines using non-manylinux docker images, where we want to build and run unit tests, they will be failed I think. Any reason you must need change this?

File: tools/ci_build/build.py [hunk 8]
```diff
@@ -1018,6 +1021,12 @@ def generate_build_tree(cmake_path, source_dir, build_dir, cuda_home, cudnn_home
     else:
         add_cmake_define_without_override(cmake_extra_defines, "onnxruntime_PYBIND_EXPORT_OPSCHEMA", "OFF")
 
+    print('build.py: 3')
+    print(cmake_extra_defines)
```
Comment (pengwa) [inline tools/ci_build/build.py:1025]: forget to remove them? :)

File: cmake/onnxruntime_training.cmake (hunk 7 above)
Comment (wschin) [reply cmake/onnxruntime_training.cmake:74]: This doesn't affect UTs, magically. From [this log](https://dev.azure.com/onnxruntime/2a773b67-e88b-4c7f-9fc0-87d31fea8ef2/_apis/build/builds/469206/logs/10), UT ran sucessfully with this change.

File: tools/ci_build/build.py (hunk 8 above)
Comment (wschin) [reply tools/ci_build/build.py:1025]: Removed. Added for investigating @snnn's [idea ](https://github.com/microsoft/onnxruntime/pull/8652#issuecomment-899882057).

File: tools/ci_build/github/azure-pipelines/templates/py-packaging-training-cuda-stage.yml [hunk 9]
```diff
@@ -128,8 +128,9 @@ stages:
                   --build_wheel \
                   --enable_onnx_tests \
                   ${{ parameters.build_py_parameters }} \
-                  --cmake_extra_defines CMAKE_CUDA_HOST_COMPILER=/opt/rh/devtoolset-$(GccVersion)/root/usr/bin/cc 'CMAKE_CUDA_ARCHITECTURES=${{ parameters.cmake_cuda_architectures }}' \
-                  --use_cuda --cuda_version=$(CudaVersion) --cuda_home=/usr/local/cuda-$(CudaVersion) --cudnn_home=/usr/local/cuda-$(CudaVersion) ;
+                  --cmake_extra_defines CMAKE_CUDA_HOST_COMPILER=/opt/rh/devtoolset-$(GccVersion)/root/usr/bin/cc 'CMAKE_CUDA_ARCHITECTURES=${{ parameters.cmake_cuda_architectures }}' onnxruntime_BUILD_UNIT_TESTS=OFF \
```
Comment (wschin) [inline tools/ci_build/github/azure-pipelines/templates/py-packaging-training-cuda-stage.yml:131]: Disable UTs for all training package pipelines so we can build packages with PythonOp.

File: tools/ci_build/github/azure-pipelines/templates/py-packaging-training-cuda-stage.yml [hunk 10]
```diff
@@ -149,7 +150,7 @@ stages:
               --volume /bert_data:/bert_data \
               --volume /hf_models_cache:/hf_models_cache \
               onnxruntimetraininggpubuild \
-                bash -c " $(PythonManylinuxDir)/bin/python3 -m pip install /build/Release/dist/$whlfilename && $(PythonManylinuxDir)/bin/python3 -m onnxruntime.training.ortmodule.torch_cpp_extensions.install && $(PythonManylinuxDir)/bin/python3 /build/Release/launch_test.py --cmd_line_with_args 'python orttraining_ortmodule_tests.py --mnist /mnist --bert_data /bert_data/hf_data/glue_data/CoLA/original/raw --transformers_cache /hf_models_cache/huggingface/transformers' --cwd /build/Release " ;
+                bash -c " $(PythonManylinuxDir)/bin/python3 -m pip install /build/Release/dist/$whlfilename && $(PythonManylinuxDir)/bin/python3 -m onnxruntime.training.ortmodule.torch_cpp_extensions.install " ;
```
Comment (wschin) [inline tools/ci_build/github/azure-pipelines/templates/py-packaging-training-cuda-stage.yml:153]: We should bring these UTs back once we isolate the Python part from ORT codebase.

**Review by snnn — COMMENTED** [review]
LGTM. Please try to get an approval from the training team.

**Review by SherlockNoMad — APPROVED** [review]

File: cmake/onnxruntime_training.cmake (hunk 7 above)
Comment (pengwa) [reply cmake/onnxruntime_training.cmake:74]: That's because all CIs have onnxruntime_ENABLE_TRAINING_TORCH_INTEROP being OFF, if you turned them on, the build will fail.

Comment (pengwa) [reply cmake/onnxruntime_training.cmake:74]: As we talked, I will change them back in my PR.
